server/gc/src/main/java/org/apache/accumulo/gc/GCRun.java

```diff
@@ -97,7 +98,8 @@ public class GCRun implements GarbageCollectionEnvironment {
   private long errors = 0;
 
   public GCRun(Ample.DataLevel level, ServerContext context) {
-    this.log = LoggerFactory.getLogger(level.name() + GCRun.class);
+    this.log = LoggerFactory.getLogger(GCRun.class.getName() + level.name());
```

Comment:
Would using a dot separator between the level and the class name make sense?  Would that benefit changing the order so that it was `LEVEL.GCRun`?  

```suggestion
    this.log = LoggerFactory.getLogger(level.name() + '.' + GCRun.class.getName());
```

Reply:
If you did that, the name would be `METADATA.org.apache.accumulo.gc.GCRun` 

I believe we want to keep the loggers under the global prefix of `org.apache.accumulo` to ensure proper [logger hierarchy](https://logging.apache.org/log4j/2.x/manual/architecture.html#logger-hierarchy). 

Having the datalevel be in front of `org.apache.accumulo` means that these log messages would not be collected if someone were to setup a general logger linked to `org.apache.accumulo`. 

---

The GC creates three GCRun objects when it starts up. Once for each datalevel (USER, ROOT, METADATA). 

The current log format layout has a precision specifier of two. So only the last two items from the logger name are shown. 

Because of the multiple instances of GCRun, and the precision specifier, if we kept the datalevel in front of  GCRun, the `.files` log entries would not be linked to their respective GCRun instance. 

```
[USER.GCRun] DEBUG: some message 1
[ROOT.GCRun] DEBUG: some message 2
[GCRun.files] DEBUG: Deleting file <hdfs://> 
[GCRun.files] DEBUG: Deleting file <hdfs://etc> 
```
Concatenating the datalevel and GCRun names ensure that the log messages can be traced back to their respective GCRun instance.  

```
[gc.GCRunUSER] DEBUG: some message 1
[gc.GCRunROOT] DEBUG: some message 2
[GCRunROOT.files] DEBUG: Deleting file <hdfs://> 
[GCRunUSER.files] DEBUG: Deleting file <hdfs://etc> 
```

We could add a dot separator after GCRun and add the datalevel afterwards. This results in log messages pointing back to a data level without the context of the class.
I don't think this would be a problem, but you lose the context of the calling class (GCRun).  
```
[GCRun.USER] DEBUG: some message 1
[GCRun.ROOT] DEBUG: some message 2
[ROOT.files] DEBUG: Deleting file <hdfs://> 
[USER.files] DEBUG: Deleting file <hdfs://etc> 
```

Reply:
Keeping the hierarchy for the general log makes sense.  For the files?  When declaring a "special" logger, you do not need to follow the strict class name hierarchy.  One example would be how we use / declare the audit log loggers, for example in AuditedSecurityOperation
```
public static final String AUDITLOG = "org.apache.accumulo.audit";
```
That lets the AUDITLOG to be called from multiple places and is grouping by functionality, not necessarily the class names.
 
So maybe something like `org.apache.accumulo.gc_files.[LEVEL]` ?  Just suggestions.

Reply:
By following the strict class name hierarchy I can ensure that the messages will be included in the normal GarbageCollector logs. 

If I were to use a logger name that's outside the current hierarchy I could end up introducing unexpected logging behavior between versions of 2.x.

Audit loggers being separate make sense as the goal is to collect common events from various classes.
If we wanted to do that for file events then I wouldn't be opposed to changing this in 3.x. 
I just don't think that's a good idea for the 2.1.x version.

I'm actually wondering if it's a better change to just reformat the deletion messages a bit to support an easy regexFilter.
I'm personally not a fan, but given the small scope (GCRun class) it's a less impactful change in the 2.1.x line.

Reply:
Maybe I have not stated my suggestion clearly?  (Objection would be to strong)  Basically I was trying to explore possibilities that (in rough priority)
   - Keep default log sufficient that it can be used for debugging - that would likely be the first place people would go look,
   - Keep the new logger separate so that it is optional - people may not want to keep / manage a separate file.
 
Overall it seems like a good change / goal.

Reply:
I modified my change to only include a log message prefix for the file actions. 
This keeps the default logger and allows users to select these logging messages via log4j or other manual means (tee + sed, grep, etc.)

Reply:
So, I believe the correct way to achieve this is using the log4j MDC/NDC feature. In log4j2 these features have been merged into one (see https://logging.apache.org/log4j/2.x/manual/thread-context.html). When using this feature your logger name stays the same and you add context to the log message via the ThreadContext. So, in this case you would add the level to the ThreadContext. in the log4j2 properties file, we would need to use `%x` in the pattern so that the ThreadContext is included in the log message in the file.

Reply:
And, I just realized that slf4j doesn't support the ThreadContext....